lib/Dialect/FIRRTL/Transforms/AssignOutputDirs.cpp

```diff
@@ -0,0 +1,108 @@
+//===- AssignOutputDirs.cpp - Assign Output Directories ---------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/Path.h"
+
+#define DEBUG_TYPE "firrtl-assign-output-dirs"
+
+using namespace circt;
+using namespace firrtl;
+namespace path = llvm::sys::path;
+
+using hw::OutputFileAttr;
+
+static StringRef lca(StringRef a, OutputFileAttr file) {
+  if (!file)
+    return StringRef();
+
+  auto b = file.getDirectory();
+  if (llvm::sys::path::is_absolute(b))
+    return StringRef();
+
+  for (auto i = path::begin(b), e = path::end(b); i != e; ++i)
+    if (*i == "..")
+      return StringRef();
+
+  size_t i = 0;
+  size_t e = std::min(a.size(), b.size());
+  for (; i < e; ++i)
+    if (a[i] != b[i])
+      break;
+
+  auto dir = a.substr(0, i);
+  if (dir.ends_with(llvm::sys::path::get_separator()))
+    return dir;
+
+  return llvm::sys::path::parent_path(dir);
+}
```

Comment:
There are a bunch of logic there that I don't think it hit in the tests. E.g., `..` and absolute path logic. I don't think we actually want this. However, everything here needs to be tested.